tests/integration/galley/webhook_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package galley
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/environment"
+	"istio.io/istio/pkg/test/framework/components/environment/kube"
+)
+
+var (
+	vwcName    = "istio-galley"
+	deployName = "istio-galley"
+	istioNs    = "istio-system"
+	sleepDelay = 3 * time.Second // How long to wait to give the reconcile loop an opportunity to act
+)
+
+var removeGalleyYaml = fmt.Sprintf(`
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: %s
+  namespace: %s
+`, deployName, istioNs)
+
+// Verify that scaling up/down doesn't modify webhook configuration
+func TestWebhookScaling(t *testing.T) {
+	framework.NewTest(t).
+		// Limit to Kube environment as we're testing integration of webhook with K8s.
+		RequiresEnvironment(environment.Kube).
+		Run(func(ctx framework.TestContext) {
+			env := ctx.Environment().(*kube.Environment)
+			startGen := getVwcGeneration(vwcName, t, env)
+
+			// Scale up
+			scaleDeployment(istioNs, deployName, 2, t, env)
+			// Wait a bit to give the ValidatingWebhookConfiguration reconcile loop an opportunity to act
+			time.Sleep(sleepDelay)
+			gen := getVwcGeneration(vwcName, t, env)
+			if gen != startGen {
+				t.Fatalf("ValidatingWebhookConfiguration was updated unexpectedly on scale up")
+			}
+
+			// Scale down
+			scaleDeployment(istioNs, deployName, 1, t, env)
+			// Wait a bit to give the ValidatingWebhookConfiguration reconcile loop an opportunity to act
+			time.Sleep(sleepDelay)
+			gen = getVwcGeneration(vwcName, t, env)
+			if gen != startGen {
+				t.Fatalf("ValidatingWebhookConfiguration was updated unexpectedly on scale down")
+			}
+		})
+}
+
+// Verify that removing the galley deployment results in the webhook configuration being removed
+func TestWebhookRemoved(t *testing.T) {
+	framework.NewTest(t).
+		// Limit to Kube environment as we're testing integration of webhook with K8s.
+		RequiresEnvironment(environment.Kube).
+		Run(func(ctx framework.TestContext) {
+			env := ctx.Environment().(*kube.Environment)
+
+			// Remove galley webhook
+			env.DeleteContents(istioNs, removeGalleyYaml)
+
+			// Verify webhook config is deleted
+			env.WaitForValidatingWebhookDeletion(vwcName)
+		})
```

Comment:
I can also think of reintroducing the deployment and observing that the config is coming back, as a test case. Another one could be deploying Galley with disable validation flag and observing that the webhook config doesn't appear.

Reply:
For the first case, what would be the best way to re-deploy whatever was originally deployed?

The second case would need to be its own test suite, correct?

Reply:
Can you scale to 0 and back up or is that different than deleting?

Reply:
It's different. The webhook configuration has ownerRefs against the galley deployment, so it gets cleaned up automatically if the deployment is removed. If the deployment is scaled to 0 but still exists the webhook remains in place.

Reply:
I did not know that, I wonder if we can use that for the injector. Sometimes I get old injector webhooks left over from old deployments that are confusing